crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -225,6 +225,116 @@ impl IndexerReader {
         Ok(stored_object)
     }
 
+    /// Fetches the past object read by object ID and version.
+    /// - If `before_version` is `false`, it looks for the exact version.
+    /// - If `true`, it finds the object read for the latest version before the
+    ///   given one.
+    pub(crate) async fn get_past_object_read_in_blocking_task(
+        &self,
+        object_id: ObjectID,
+        version: SequenceNumber,
+        before_version: bool,
+    ) -> Result<PastObjectRead, IndexerError> {
+        let (history_object, latest_version) = self
+            .spawn_blocking(move |this| this.query_past_object(&object_id, version, before_version))
+            .await?;
+
+        match (history_object, latest_version) {
+            (Some(object), _) => {
+                object
+                    .try_into_past_object_read(self.package_resolver.clone())
+                    .await
+            }
+
+            // If the object never existed
+            (None, None) => Ok(PastObjectRead::ObjectNotExists(object_id)),
+
+            // If the object exists, but the requested version is too high
+            (None, Some(latest)) if version.value() > latest as u64 => {
+                Ok(PastObjectRead::VersionTooHigh {
+                    object_id,
+                    asked_version: version,
+                    latest_version: SequenceNumber::from(latest as u64),
+                })
+            }
+
+            // If the object exists but not at this version
+            (None, Some(_)) => Ok(PastObjectRead::VersionNotFound(object_id, version)),
+        }
+    }
+
+    /// Fetches a past object by its ID and version.
+    ///
+    /// - If `before_version` is `false`, it looks for the exact version.
+    /// - If `true`, it finds the latest version before the given one.
+    ///
+    /// Also retrieves the latest available version of the object to check if:
+    /// - The requested version is too high.
+    /// - The object exists but not at this version.
+    /// - The object doesn’t exist at all.
+    ///
+    /// Searches `objects_history`.
+    fn query_past_object(
+        &self,
+        object_id: &ObjectID,
+        version: SequenceNumber,
+        before_version: bool,
+    ) -> Result<(Option<StoredHistoryObject>, Option<i64>), IndexerError> {
+        let object_id_vec = object_id.to_vec();
+        let version_num = version.value() as i64;
+
+        // Query objects_history for the latest known version and checkpoint
+        let latest_history_object_opt: Option<StoredHistoryObject> =
+            run_query!(&self.pool, |conn| {
+                objects_history::dsl::objects_history
+                    .filter(objects_history::object_id.eq(&object_id_vec))
+                    .order_by(objects_history::object_version.desc())
+                    .limit(1)
+                    .first::<StoredHistoryObject>(conn)
+                    .optional()
+            })?;
```

Comment:
If we must execute two queries in most cases, perhaps you can get the max version from the `objects_version` table, for efficiency.

Reply:
The `objects_version` table has a primary key on (object_id, object_version) without other indexes, so this would also not be a very efficient query I assume. Also, in case we add pruning, we would need to take care and consider this additional table too, which adds further complexity.

Reply:
But here you query on `object_id` and order by `version` which should use the `(object_id, object_version)` index.

Reply:
Fixed in https://github.com/iotaledger/iota/commit/c3d2b07d4e1216ffafef45e96b86418b2162b75b